### ISSUE TITLE
`private_interfaces` lint needs to perform type resolution to avoid false positives

### ISSUE THREAD

**Issue by mqudsi** [rust-lang/rust#149284]
The `private_interfaces` lint incorrectly fires for cases where a type is declared *via* a private interface but nevertheless fully resolves to an equally (or more) public type.

```rust
trait Service {
    type Output;
    
    fn process() -> Self::Output;
}

pub struct SomeService;
impl Service for SomeService {
    type Output = ();
    
    fn process() -> Self::Output {
        todo!();
    }
}

// False positive on return type here:
pub fn exposed_method() -> <SomeService as Service>::Output {
    SomeService::process()
}
```

Currently produces the following error on both 1.91.1 stable and 1.93.0-nightly:

```
warning: associated type `Service::Output` is more private than the item `exposed_method`
  --> src/lib.rs:16:1
   |
16 | pub fn exposed_method() -> <SomeService as Service>::Output {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ function `exposed_method` is reachable at visibility `pub`
   |
note: but associated type `Service::Output` is only usable at visibility `pub(crate)`
  --> src/lib.rs:2:5
   |
 2 |     type Output;
   |     ^^^^^^^^^^^
   = note: `#[warn(private_interfaces)]` on by default

warning: trait `Service` is more private than the item `exposed_method`
  --> src/lib.rs:16:1
   |
16 | pub fn exposed_method() -> <SomeService as Service>::Output {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ function `exposed_method` is reachable at visibility `pub`
   |
note: but trait `Service` is only usable at visibility `pub(crate)`
  --> src/lib.rs:1:1
   |
 1 | trait Service {
   | ^^^^^^^^^^^^^
```

Playground link: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=4756e3d350f5f0bdd90086f76d043a89

As you can see, the `private_interfaces` lint is automatically fired when a less-public type is encountered in the type signature, but if type resolution were performed it would be obvious that this actually resolves to a very much public and accessible type (in this case, `()`).

(This is a very much contrived example; in my actual code I have a trait that exposes blocking methods via a non-blocking async interface and the trait's associated type is the shortest and least error-prone method of expressing the type of a `pub static`)

@rustbot label +T-compiler +L-private_interfaces +L-false-positive +A-visibility +A-lints +C-bug
